src/Raven.Server/Documents/PeriodicBackup/DirectUpload/DirectFileUploader.cs

```diff
@@ -0,0 +1,199 @@
+using Raven.Server.Documents.PeriodicBackup.Retention;
+using Raven.Server.ServerWide;
+using Raven.Server.Utils;
+using Sparrow.Logging;
+using Sparrow.Server.Logging;
+using Sparrow.Server.Utils;
+
+namespace Raven.Server.Documents.PeriodicBackup.DirectUpload;
+
+public sealed class DirectFileUploader : FileUploaderBase
+{
+    private readonly short _concurrentThreads;
+    private readonly BackupConfiguration.BackupDestination _destination;
+
+    public DirectFileUploader(UploaderSettings settings, RetentionPolicyBaseParameters retentionPolicyParameters, RavenLogger logger, BackupResult backupResult, Action<IOperationProgress> onProgress, OperationCancelToken taskCancelToken) :
+        base(settings, retentionPolicyParameters, logger, backupResult, onProgress, taskCancelToken)
+    {
+        _destination = settings.Destination;
+        _concurrentThreads = settings.ConcurrentThreads;
+    }
+
+    internal Stream StreamForBackupDestination(DocumentDatabase database, string folderName, string fileName)
+    {
+        switch (_destination)
+        {
+            case BackupConfiguration.BackupDestination.AmazonS3:
+                return new AwsS3DirectUploadStream(GetDirectUploadParameters(
+                    progress => new RavenAwsS3Client(_settings.S3Settings, database.Configuration.Backup, progress, TaskCancelToken.Token),
+                    _settings.S3Settings.RemoteFolderName, folderName, fileName));
+
+            case BackupConfiguration.BackupDestination.Azure:
+                return new AzureDirectUploadStream(GetDirectUploadParameters(
+                    progress => RavenAzureClient.Create(_settings.AzureSettings, database.Configuration.Backup, progress, TaskCancelToken.Token),
+                    _settings.AzureSettings.RemoteFolderName, folderName, fileName));
+
+            default:
+                throw new ArgumentOutOfRangeException($"Missing implementation for direct upload destination '{_destination}'");
+        }
+    }
+
+    private DirectUploadStream<T>.Parameters GetDirectUploadParameters<T>(Func<Progress, T> clientFactory, string remoteFolderName, string folderName, string fileName) where T : IDirectUploader
+    {
+        return new DirectUploadStream<T>.Parameters
+        {
+            ClientFactory = clientFactory,
+            Key = CombinePathAndKey(remoteFolderName, folderName, fileName),
+            Metadata = new Dictionary<string, string>
+            {
+                { "Description", GetBackupDescription() }
+            },
+            IsFullBackup = _isFullBackup,
+            RetentionPolicyParameters = _retentionPolicyParameters,
+            CloudUploadStatus = _backupResult.S3Backup,
+            OnBackupException = _settings.OnBackupException,
+            OnProgress = AddInfo
+        };
+    }
+
+    public override string GetBackupDescription()
+    {
+        return $"{nameof(DirectFileUploader)}";
+    }
+
+    public void AddDelete(string folderName, string fileName)
```

Comment:
Why are you using threads here instead of tasks?

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-24606/Retire-attachments-Server-Side-Use-Task-instead-of-Thread-in-DirectFileUploader.cs-Phase-1